pubsub/pubsub.go

```diff
@@ -0,0 +1,272 @@
+	"google.golang.org/api/support/bundler"
+)
+
+// Message contains data to be published.
+type Message struct {
+	// Body contains the content of the message.
+	Body []byte
+
+	// Attributes has key/value metadata for the message.
+	Attributes map[string]string
+
+	// AckID identifies the message on the server.
+	// It can be used to ack the message after it has been received.
+	ackID AckID
+
+	// sub is the Subscription this message was received from.
+	sub *Subscription
+}
+
+// AckID identifies a message for acknowledgement.
+type AckID interface{}
+
+// Ack acknowledges the message, telling the server that it does not need to
+// be sent again to the associated Subscription. This method blocks until
+// the message has been confirmed as acknowledged on the server, or failure
+// occurs.
+func (m *Message) Ack(ctx context.Context) error {
+	// Send the message back to the subscription for ack batching.
+	mec := msgErrChan{
+		msg:     m,
+		errChan: make(chan error),
+	}
+	size := 8
+	if err := m.sub.ackBatcher.AddWait(ctx, mec, size); err != nil {
+		return err
+	}
+	return <-mec.errChan
+}
+
+type msgErrChan struct {
+	msg     *Message
+	errChan chan error
+}
+
+// Topic publishes messages to all its subscribers.
+type Topic struct {
+	driver  driver.Topic
+	batcher *bundler.Bundler
+}
+
+// TopicOptions contains configuration for Topics.
+type TopicOptions struct {
+	// SendDelay tells the max duration to wait before sending the next batch of
+	// messages to the server.
+	SendDelay time.Duration
+
+	// BatchSize specifies the maximum number of messages that can go in a batch
+	// for sending.
+	BatchSize int
+}
+
+// SendDelayDefault is the value for TopicOptions.SendDelay if it is not set.
+const SendDelayDefault = time.Millisecond
+
+// Send publishes a message. It only returns after the message has been
+// sent, or failed to be sent. Send can be called from multiple goroutines
+// at once.
+func (t *Topic) Send(ctx context.Context, m *Message) error {
+	mec := msgErrChan{
+		msg:     m,
+		errChan: make(chan error),
+	}
+	size := len(m.Body)
+	for k, v := range m.Attributes {
+		size += len(k)
+		size += len(v)
+	}
+	if err := t.batcher.AddWait(ctx, mec, size); err != nil {
+		return err
+	}
+	return <-mec.errChan
+}
+
+// Close flushes pending message sends and disconnects the Topic.
+// It only returns after all pending messages have been sent.
+func (t *Topic) Close() error {
+	t.batcher.Flush()
+	return t.driver.Close()
+}
+
+// NewTopic makes a pubsub.Topic from a driver.Topic and opts to
+// tune how messages are sent. Behind the scenes, NewTopic spins up a goroutine
+// to bundle messages into batches and send them to the server.
+func NewTopic(ctx context.Context, d driver.Topic, opts TopicOptions) *Topic {
+	handler := func(item interface{}) {
+		mecs, ok := item.([]msgErrChan)
+		if !ok {
+			panic("failed conversion to []msgErrChan in bundler handler")
+		}
+		var dms []*driver.Message
```

Comment:
You can initialize dms to the right size, e.g.:

```
dms := make([]*driver.Message, len(mecs))
```
and then

```
for i, mec := range mecs {
  dms[i] = ...
```

Reply:
I did a similar thing earlier with 
```go
dms := make([]*driver.Message, 0, len(mecs))
...
for i, mec := range mecs {
  ...
  dms = append(dms, dm)
}
```
and Jonathan said it was a premature optimization and asked me to put it the way it is now.

I'm making this change, but would like to leave this code alone after this.